File: ch2/sort/src/bitonic.ts
// bitonic sort

// [4, 5, 6, 8]
const main = (arr: number[]) => {
  // 4, 2, 1
  const sort = (arr: number[], up: boolean): number[] => {
    if (arr.length <= 1) {
      return arr;
    }

    const mid = Math.floor(arr.length / 2);
    // 2, 1, 2 (4,5)
    const left = sort(arr.slice(0, mid), true);
    // 2, 1, 2 (6,8)
    const right = sort(arr.slice(mid), false);

    // 1 + 1 (5,4), 1 + 1(7,8), 2 + 2 (4,5,6,8)
    const merged = left.concat(right);

    return subSort(merged, up);
  };

  // ここを並列化するのがバイトニックソートの特徴、この比較は他の処理に依存しないので並列化できる
  // まず半分に割って、その後最小単位（2）まで分割してソート、その後マージしていく
  const subSort = (arr: number[], up: boolean): number[] => {
    if (arr.length <= 1) {
      return arr;
    }
    const mid = Math.floor(arr.length / 2);

    const swappedArr = compareAndSwap(arr, up);

    const left = swappedArr.slice(0, mid);
    const right = swappedArr.slice(mid);

    const sortedLeft = subSort(left, up);
    const sortedRight = subSort(right, up);

    return sortedLeft.concat(sortedRight);
  };

  const compareAndSwap = (arr: number[], up: boolean): number[] => {
    const mid = Math.floor(arr.length / 2);
    for (let i = 0; i < mid; i++) {
      // (5,4), (4,5,6,8)
      if (arr[i] > arr[mid + i] === up) {
        const temp = arr[i];
        arr[i] = arr[mid + i];
        arr[mid + i] = temp;
        // (4,5), (4,5,6,8)
      }
    }
    return arr;
  };

  const res = sort(arr, true);
  console.log(res);
};

main([5, 4]); // [1, 2, 3, 4, 5, 6, 7, 8]

File: ch2/sort/src/bitonic.py
def is_power_of_two(n):
    return (n != 0) and (n & (n - 1) == 0)


def main(arr: list[int]):
    if not is_power_of_two(len(arr)):
        print("Array is not a power of 2")
    if len(arr) <= 1:
        return arr

    print(f"sorting {arr}")

    def sort(arr: list[int], up: bool):
        print(f"sorting {arr} in {up} order")
        if len(arr) <= 1:
            return arr
        else:
            mid = len(arr) // 2
            first = sort(arr[mid:], True)
            second = sort(arr[:mid], False)
            mergedArr = first + second
            return sub_sort(mergedArr, up)

    def sub_sort(arr: list[int], up: bool):
        print(f"sub_sorting {arr} in {up} order")
        if len(arr) <= 1:
            return arr
        else:
            mid = len(arr) // 2
            compare_and_swap(arr, up)
            first = sub_sort(arr[:mid], up)
            second = sub_sort(arr[mid:], up)
            return first + second

    def compare_and_swap(arr: list[int], up: bool):
        mid = len(arr) // 2
        for i in range(mid):
            if (arr[i] > arr[mid + i]) == up:
                arr[i], arr[mid + i] = arr[mid + i], arr[i]

    return sort(arr, True)


if __name__ == "__main__":
    result = main([3, 7, 4, 8, 6, 2, 1, 5, 9, 10, 12, 11, 13, 15, 14, 16])
    print(result)
